File: c/medium/smallest_subtree_with_all_the_deepest_nodes.c
#include <stdlib.h>

typedef struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

int global_max_depth = 0;

typedef struct {
    TreeNode* node;
    int depth;
} Result;

Result dfs(TreeNode* node, int current_depth) {
    if (node == NULL) {
        if (current_depth > global_max_depth)
            global_max_depth = current_depth;
        Result r = { NULL, current_depth };
        return r;
    }

    Result left = dfs(node->left, current_depth + 1);
    Result right = dfs(node->right, current_depth + 1);

    if (left.depth == right.depth && left.depth == global_max_depth) {
        Result r = { node, left.depth };
        return r;
    }

    return (left.depth > right.depth) ? left : right;
}

TreeNode* subtreeWithAllDeepest(TreeNode* root) {
    global_max_depth = 0;
    Result r = dfs(root, 0);
    return r.node;
}

File: rust/medium/smallest_subtree_with_all_the_deepest_nodes.rs
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

pub struct Solution {
    global_max_depth: i32,
}

impl Solution {
    pub fn subtree_with_all_deepest(
        root: Option<Rc<RefCell<TreeNode>>>
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let mut s = Solution { global_max_depth: 0 };
        let (node, _) = s.dfs(root, 0);
        node
    }

    fn dfs(
        &mut self,
        node: Option<Rc<RefCell<TreeNode>>>,
        current_depth: i32,
    ) -> (Option<Rc<RefCell<TreeNode>>>, i32) {
        match node {
            None => {
                self.global_max_depth = self.global_max_depth.max(current_depth);
                (None, current_depth)
            }
            Some(n) => {
                let left = self.dfs(n.borrow().left.clone(), current_depth + 1);
                let right = self.dfs(n.borrow().right.clone(), current_depth + 1);

                if left.1 == right.1 && left.1 == self.global_max_depth {
                    (Some(n), left.1)
                } else if left.1 > right.1 {
                    left
                } else {
                    right
                }
            }
        }
    }
}
